File: Leetcode_All_Solved/146.cpp
#include<list>
#include<unordered_map>
#include<utility>
class LRUCache {
private:
    int capacity;
    list<pair<int,int>> dll;
    unordered_map<int, list<pair<int,int>>::iterator > key2dll_iter;
    inline void move_front(list<pair<int,int>>::iterator it){
        dll.splice(dll.begin(), dll, it);//NOTE: stl function faster
        return;
        
        /*dll.push_front( {it->first, it->second} ); //(*it);//copy or move
        int key = it->first;
        dll.erase(it);
        key2dll_iter[key] = dll.begin();*/
    }
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }
    
    int get(int key) {
        auto f = key2dll_iter.find(key);
        if(f == key2dll_iter.end())return (-1);
        
        //move to front
        int val = f->second->second;//list iter destroyed
        move_front(f->second);
        return val;
    }
    
    void put(int key, int value) {
        auto f = key2dll_iter.find(key);
        if(f != key2dll_iter.end()){
            f->second->second = value;
            move_front(f->second);
            return;
        }
        
        if(dll.size() == capacity){
            int back_key = dll.back().first;
            dll.pop_back();
            key2dll_iter.erase(back_key);
        }
        dll.push_front( {key,value} );
        key2dll_iter[key] = dll.begin();
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: Leetcode_All_Solved/146.py
class Node:
    def __init__(self, key=(-1), value=(-1)):
        self.key = key
        self.value = value
        self.next = None
        self.prev = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0

        # dummy nodes
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

        self.key2node = dict()

    def get(self, key: int) -> int:
        if(key not in self.key2node):
            return (-1)
        
        cur = self.key2node[key]
        self.move_front(cur)
        return cur.value

    def put(self, key: int, value: int) -> None:
        if(key not in self.key2node):
            node = Node(key, value)
            self.push_front(node)

            if(self.size > self.capacity):
                self.pop_tail()
        else:
            node = self.key2node[key]
            node.value = value
            self.move_front(node)
    
    def push_front(self, node):
        first = self.head.next

        node.next = first
        first.prev = node

        self.head.next = node
        node.prev = self.head

        self.size += 1
        self.key2node[node.key] = node
    
    def move_front(self, node):
        self.pop_node(node)
        self.push_front(node)

    def pop_node(self, node):
        prev = node.prev
        next = node.next

        prev.next = next
        next.prev = prev

        self.size -= 1
        assert node.key in self.key2node, f'key {node.key} not found!!'
        self.key2node.pop(node.key)

    def pop_tail(self):
        self.pop_node(self.tail.prev)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
